### PR TITLE
fix: [CDS-45488]: Move successful zombie monitor behind FF WORKFLOW_EXECUTION_ZOMBIE_MONITOR

### PR DESCRIPTION
**Pull request by Fernando-Dourado**
## Describe your changes
- Add check if the parent state execution instance is a fork
- Change status check to verify for success only

## Context
We released a feature to kill zombie workflows in case of manager restart, we enabled it globally weeks ago. Then we pushed changes to cover one more case because of another issue, this change was a wrong condition that detect a skipped step from a fork as a zombie and trigger the execution abort.

A zombie is a step inside a workflow where the transition between steps fails and leaves them in an incorrect condition, most of the time it leaves in running status without being running. The changes we pushed were to cover a case where the workflow keeps the running status, but all the steps already reach the success status.

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/39225)

### CODE REVIEW

**Comment by Fernando-Dourado** [comment]
trigger checkstyle, pmd

**Comment by Fernando-Dourado** [comment]
trigger messagecheck

**Review by inderschera — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39225)* status: 2 file(s) to review. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java` line 48 at r1](https://reviewable.io/reviews/harness/harness-core/39225#-NFOgeGz75xbFQ2Y2tIy:-NFOgeGz75xbFQ2Y2tIz:bi36hpk) ([raw file](https://github.com/harness/harness-core/blob/6bf7f688e6eee73ec2050ac8cf648860bba3e372/400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java#L48)):*
> ```Java
> @Slf4j
> @TargetModule(HarnessModule._815_CG_TRIGGERS)
> public class WorkflowExecutionZombieHandler implements MongoPersistenceIterator.Handler<WorkflowExecution> {
> ```

Since it's not an iterator, should we remove implements MongoPersistenceIterator.Handler<WorkflowExecution> ?
___
*[`400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java` line 60 at r1](https://reviewable.io/reviews/harness/harness-core/39225#-NFOfzJa3vK3co4RrkZB:-NFOfzJa3vK3co4RrkZC:b-20ul35) ([raw file](https://github.com/harness/harness-core/blob/e19576ff44df3e7fc4019ce82286a1ff08459976/400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java#L60)):*
> ```Java
>   @Override
>   public void handle(WorkflowExecution wfExecution) {
>     if (featureFlagService.isNotEnabled(FeatureName.WORKFLOW_EXECUTION_ZOMBIE_MONITOR, wfExecution.getAccountId())) {
> ```

why have we moved feature flag check down?
___
*[`400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java` line 137 at r1](https://reviewable.io/reviews/harness/harness-core/39225#-NFOitbe27DBn4_V9FtA:-NFOitbe27DBn4_V9FtB:bruc7pt) ([raw file](https://github.com/harness/harness-core/blob/6bf7f688e6eee73ec2050ac8cf648860bba3e372/400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java#L137)):*
> ```Java
>       return false;
>     }
>     return hitEndedThreshold(seInstance) && ExecutionStatus.SUCCESS.equals(seInstance.getStatus())
> ```

Why not ExecutionStatus.isPositiveStatus(seInstance.getStatus()) here?
___
*[`400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java` line 138 at r1](https://reviewable.io/reviews/harness/harness-core/39225#-NFOj7lt04rpy_DWnVr7:-NFOj7lt04rpy_DWnVr8:bwk9bcd) ([raw file](https://github.com/harness/harness-core/blob/6bf7f688e6eee73ec2050ac8cf648860bba3e372/400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java#L138)):*
> ```Java
>     }
>     return hitEndedThreshold(seInstance) && ExecutionStatus.SUCCESS.equals(seInstance.getStatus())
>         && isNotParentFork(seInstance);
> ```

What if parent was a fork and all it's children had completed but workflow execution didn't proceed after that? Should we handle that case as well?

**Review by Fernando-Dourado — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39225)* status: 2 file(s) to review. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java` line 48 at r1](https://reviewable.io/reviews/harness/harness-core/39225#-NFOgeGz75xbFQ2Y2tIy:-NFPgoxo5qMebfqaLBF7:b-4luen6) ([raw file](https://github.com/harness/harness-core/blob/6bf7f688e6eee73ec2050ac8cf648860bba3e372/400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java#L48)):*
<details><summary><i>Previously, inderschera (Inderpreet Chera) wrote…</i></summary><blockquote>

Since it's not an iterator, should we remove implements MongoPersistenceIterator.Handler<WorkflowExecution> ?
</blockquote></details>

Yes, you are right here. It is a miss from the previous review when the class does not need to implement this interface anymore.
___
*[`400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java` line 60 at r1](https://reviewable.io/reviews/harness/harness-core/39225#-NFOfzJa3vK3co4RrkZB:-NFPpy9v4iLzvLcUv6Jy:b-a8uytk) ([raw file](https://github.com/harness/harness-core/blob/e19576ff44df3e7fc4019ce82286a1ff08459976/400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java#L60)):*
<details><summary><i>Previously, inderschera (Inderpreet Chera) wrote…</i></summary><blockquote>

why have we moved feature flag check down?
</blockquote></details>

The FF was on for a long time in PROD, then in our last release we made a change to cover one additional case which we named "success zombie". Then we got a regression and started to abort not zombie executions.

We decide to keep the FF only to this new condition about success zombie and not the already GA code of zombie handler.
___
*[`400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java` line 137 at r1](https://reviewable.io/reviews/harness/harness-core/39225#-NFOitbe27DBn4_V9FtA:-NFPrPVT1jIW9qJhZ2Ag:b-syk76l) ([raw file](https://github.com/harness/harness-core/blob/6bf7f688e6eee73ec2050ac8cf648860bba3e372/400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java#L137)):*
<details><summary><i>Previously, inderschera (Inderpreet Chera) wrote…</i></summary><blockquote>

Why not ExecutionStatus.isPositiveStatus(seInstance.getStatus()) here?
</blockquote></details>

The main reason is to avoid additional side effects, there are many cases and conditions to detect a stuck execution. We choose to be more restrictive and specific instead of more open, as we doing a rework to zombie and pod restart in cg-manager.
___
*[`400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java` line 138 at r1](https://reviewable.io/reviews/harness/harness-core/39225#-NFOj7lt04rpy_DWnVr7:-NFPrvbhDMx6cpYU4tS1:bvu2kc5) ([raw file](https://github.com/harness/harness-core/blob/6bf7f688e6eee73ec2050ac8cf648860bba3e372/400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java#L138)):*
<details><summary><i>Previously, inderschera (Inderpreet Chera) wrote…</i></summary><blockquote>

What if parent was a fork and all it's children had completed but workflow execution didn't proceed after that? Should we handle that case as well?
</blockquote></details>

Not now, this is one of the other cases we are not addressing for now.

**Review by inderschera — APPROVED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39225)* status: 1. unresolved discussions.

File: 400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java [hunk 1]
```diff
@@ -126,13 +124,29 @@ private boolean isZombie(StateExecutionInstance seInstance) {
...
   private boolean isSuccessZombie(StateExecutionInstance seInstance) {
-    return hitEndedThreshold(seInstance) && ExecutionStatus.isPositiveStatus(seInstance.getStatus());
+    if (featureFlagService.isNotEnabled(FeatureName.WORKFLOW_EXECUTION_ZOMBIE_MONITOR, seInstance.getAccountId())) {
+      return false;
+    }
+    return hitEndedThreshold(seInstance) && ExecutionStatus.SUCCESS.equals(seInstance.getStatus())
+        && isNotParentFork(seInstance);
+  }
+
+  private boolean isNotParentFork(StateExecutionInstance seInstance) {
```
Comment (lasdox) [inline 400-rest/src/main/java/io/harness/workers/background/critical/iterator/WorkflowExecutionZombieHandler.java:141]: I suggest to change this method name to: "hasNoForkParent" but this is just preference

**Review by lasdox — APPROVED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39225)* status: 1. unresolved discussions.

**Review by vcalasansh — APPROVED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39225)* status: 1. unresolved discussions.

**Pull request merged**